### PR TITLE
OHARA-1597 StreamApp Action response should be idempotent

### PR DESCRIPTION
**Pull request by saivirtue**

### CODE REVIEW

**Comment by saivirtue** [comment]
I will add more tests soon

**Comment by chia7712** [comment]
Why we don't check the existence of stream app first?

File: ohara-client/src/main/scala/com/island/ohara/client/configurator/v0/StreamApi.scala [hunk 1]
```diff
@@ -123,6 +125,7 @@ object StreamApi {
                              fromTopics: Seq[String],
                              toTopics: Seq[String],
                              lastModified: Long,
+                             state: Option[ConnectorState] = None,
```
[the remark this answers is not in the document]
Comment (saivirtue) [reply ohara-client/src/main/scala/com/island/ohara/client/configurator/v0/StreamApi.scala]: OK, have created.
- [ ] OHARA-1603

File: ohara-configurator/src/main/scala/com/island/ohara/configurator/route/StreamRoute.scala [hunk 2]
```diff
@@ -253,103 +280,115 @@ private[configurator] object StreamRoute {
...
-                              id,
-                              Some(ContainerApi.ContainerState.EXITED)
-                            )
+                    }
+                    .map {
+                      case (client, brokerCluster) =>
+                        brokerCollie.cluster(brokerCluster).map {
+                          case (brokerInfo, _) =>
+                            val brokerList =
+                              brokerInfo.nodeNames.map(_ + ":9092").mkString(",")
```
[the remark this answers is not in the document]
Comment (saivirtue) [reply ohara-configurator/src/main/scala/com/island/ohara/configurator/route/StreamRoute.scala]: hmm...lose this constant string, thanks!

**Comment by saivirtue** [comment]
> Why we don't check the existence of stream app first?

streamApp may be changed the state after a period, so I think the streamApp info should be queried again to get the newest state.

**Comment by chia7712** [comment]
Could you add test for idempotent operation?

**Comment by chia7712** [comment]
我對現在start streamapp的程式碼認知如下：

1. 透過id取得streamapp data
1. 檢查參數
1. 隨機挑選節點
1. 更新streamapp的運作節點資訊
1. 取得dockerclient
1. 檢查topics並取得broker cluster name
1. 建立broker client
1. 取得jars並且啟動streamapp
1. 建立response

在我想像裡面，這邊要做的idempotent應該是start streamapp應該do nothing當該streamapp"已經"在執行了，此時就是回傳該streamapp的"最新狀態"且不用去真的"start streamapp"。但我從目前的代碼裡面無法看出上述邏輯

麻煩幫忙看一下上述是否有錯，謝謝

**Comment by saivirtue** [comment]
> 我對現在start streamapp的程式碼認知如下：
> 
> 1. 透過id取得streamapp data
> 2. 檢查參數
> 3. 隨機挑選節點
> 4. 更新streamapp的運作節點資訊
> 5. 取得dockerclient
> 6. 檢查topics並取得broker cluster name
> 7. 建立broker client
> 8. 取得jars並且啟動streamapp
> 9. 建立response
> 
> 在我想像裡面，這邊要做的idempotent應該是start streamapp應該do nothing當該streamapp"已經"在執行了，此時就是回傳該streamapp的"最新狀態"且不用去真的"start streamapp"。但我從目前的代碼裡面無法看出上述邏輯
> 
> 麻煩幫忙看一下上述是否有錯，謝謝

The steps are right except 8.
when starting the streamApp, first will check if there is an running container (by filter appId) and get the container state again. It should be idempotent if the streamApp is running normal or already exited.
for example,
first time : /start => api response { id: "id", state : running } (always running if there is no exception)
second and futher time : /start => api response { id : "id", state : running / exited } (base on the query result of the container)

**Comment by saivirtue** [comment]
I think we should support some "state query" operation if the UI has following operation:
1. add streamApp and click start button
2. api response running
3. after a while, UI refresh or click the start button again
4. api should response the "current state" of the streamApp

**Comment by chia7712** [comment]
> when starting the streamApp, first will check if there is an running container (by filter appId) and get the container state again. It should be idempotent if the streamApp is running normal or already exited.
for example,

Got it. Thanks!

**Comment by saivirtue** [comment]
> Could you add test for idempotent operation?

Currently there is only a idempotent test for fail scenario...maybe add more "real" case in ohara-it module

- [ ] #326

**Comment by chia7712** [comment]
> Currently there is only a idempotent test for fail scenario...maybe add more "real" case in ohara-it module

It would be better to test it on UT... Our server will burn out is we put too many tests on IT :)

**Comment by saivirtue** [comment]
> > Currently there is only a idempotent test for fail scenario...maybe add more "real" case in ohara-it module
> 
> It would be better to test it on UT... Our server will burn out is we put too many tests on IT :)

XD. I don't have much thought about how to dynamic "package a custom streamApp jar" into UT test...do you have any good idea or just add a "pre-defined" custom streamApp jar?

**Comment by chia7712** [comment]
bq. do you have any good idea or just add a "pre-defined" custom streamApp jar?

In ut we don't need to run a streamapp actually...Our point is "idempotent operation".

Probably streamroute needs some refactor to address this comment. You can address it on follow-up.

**Comment by saivirtue** [comment]
> bq. do you have any good idea or just add a "pre-defined" custom streamApp jar?
> 
> In ut we don't need to run a streamapp actually...Our point is "idempotent operation".
> 
> Probably streamroute needs some refactor to address this comment. You can address it on follow-up.

OK, thanks a ton!
I will add some fixes base on previous comment.
1. replace :9092
2. check the streamApp running node exists or not before get random node

**Comment by chia7712** [comment]
another Q.

Is "a exited streamapp" equal with "removed streamapp"? For example, do we create another streamapp if there is already a exited streamapp (of course, them have identical id).

**Comment by chia7712** [comment]
> nope. And for that case, start action will do nothing but only get the exited streamApp status. User must stop (a.k.a. remove streamApp) streamApp first to start the same streamApp again. id is generated at the time that streamApp add in the pipeline

Got it. +1 to this PR

**Comment by saivirtue** [comment]
retry

**Pull request merged**